### PR TITLE
Tx tables fix

### PR DESCRIPTION
**Pull request by undertome**
This is just an extension for a previously approved PR - #3815 which is from a deleted repository.

I added some changes in an attempt to improve logic and readbility. Fortunately these changes are included in a separate commit to make things easier.

### CODE REVIEW

File: src/ripple/app/rdb/RelationalDBInterface_nodes.h [hunk 1]
```diff
@@ -277,7 +277,11 @@ getTxHistory(
  * @param j Journal.
  * @return Vector of pairs of found transactions and their metadata
  *         sorted in ascending order by account sequence.
- *         Also number of transactions processed.
+ *         Also number of transactions processed or skpiied.
```
Comment (cjcobb23) [inline src/ripple/app/rdb/RelationalDBInterface_nodes.h:280]: typo on `skpiied`

File: src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp [hunk 2]
```diff
@@ -1175,32 +1351,38 @@ RelationalDBInterfaceSqliteImp::oldestAccountTxPage(
...
-                limit_used,
-                page_length);
-            opt.marker = marker;
-            if (total < 0)
-                return false;
-            limit_used += total;
-            return true;
-        });
+    /* else use shard databases, if available */
+    if (shardStoreExists() && useTxTables_)
```
Comment (cjcobb23) [inline src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp:1355]: I think all of these functions can just return early if `useTxTables_` is false. There's no need to do anything else. So at the top of the function:

```
if(!useTxTables_)
    return {};
```

**Review by cjcobb23 — COMMENTED** [review]
There are a lot of places where we have to check the value of `useTxTables_`. It's seems like this might be a little error prone. What do you think of creating two separate classes, one for when `useTxTables_` is true, and one for when it's false? The class used in the false case would just return an empty optional for all of the functions that query the tx tables.

File: src/ripple/app/rdb/RelationalDBInterface_nodes.h [hunk 3]
```diff
@@ -277,7 +277,11 @@ getTxHistory(
...
+ *         If this number is >= 0, then it means number of transactions
+ *         processed, if it is < 0, then ~number means number of transactions
+ *         skipped. We need to skip some quantity of transactions if option
```
Comment (seelabs) [inline src/ripple/app/rdb/RelationalDBInterface_nodes.h:283]: Returning `~number` for the number of transactions skipped is a very strange interface. I'd suggest returning `-number` instead - unless there's some strange reason why `~number` is better, in which case we should document that reason (here and other functions as well).

File: src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp [hunk 4]
```diff
@@ -457,14 +470,20 @@ RelationalDBInterfaceSqliteImp::getMinLedgerSeq()
...
 
-    /* else use shard databases */
-    std::optional<LedgerIndex> res;
-    iterateLedgerForward(
-        {}, [&](soci::session& session, std::uint32_t shardIndex) {
-            res = ripple::getMinLedgerSeq(session, TableType::Ledgers);
-            return !res;
-        });
-    return res;
+    /* else use shard databases, if available */
```
Comment (seelabs) [inline src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp:473]: The code is better off without comments that repeat what the next line of code does (here and in many other places in this file). You don't need to change that now (it's a minor nit), but new code should not include such comments.

Comment (undertome) [reply src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp:473]: Noted. I'll consider removing these as I dig deeper into refactoring.

File: src/ripple/app/rdb/RelationalDBInterface_nodes.h (hunk 1 above)
Comment (undertome) [reply src/ripple/app/rdb/RelationalDBInterface_nodes.h:280]: Done

File: src/ripple/app/rdb/RelationalDBInterface_nodes.h (hunk 3 above)
Comment (undertome) [reply src/ripple/app/rdb/RelationalDBInterface_nodes.h:283]: Done

File: src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp [hunk 5]
```diff
@@ -257,23 +260,33 @@ class RelationalDBInterfaceSqliteImp : public RelationalDBInterfaceSqlite
...
     /**
-     * @brief checkoutLedger Checks if node ledger DB exists.
+     * @brief existsLedger Checks if node ledger DB exists.
      * @return True if node ledger DB exists.
      */
     bool
     existsLedger()
     {
-        return !!lgrdb_;
+        return (bool)lgrdb_;
```
Comment (mtrippled) [inline src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp:269]: the bool() operator automatically evaluates whether a unique_ptr contains something, so casting is redundant.

Also, the way to cast this if you really needed to would be by using something like static_cast<bool>(anotherpodtype);

File: src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp [hunk 6]
```diff
@@ -257,23 +260,33 @@ class RelationalDBInterfaceSqliteImp : public RelationalDBInterfaceSqlite
...
     /**
-     * @brief checkoutTransaction Checks if node transaction DB exists.
+     * @brief existsTransaction Checks if node transaction DB exists.
      * @return True if node transaction DB exists.
      */
     bool
     existsTransaction()
     {
-        return !!txdb_;
+        return (bool)txdb_;
```
Comment (mtrippled) [inline src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp:279]: See casting comment above.

**Review by mtrippled — CHANGES_REQUESTED** [review]
I tested that this works with use_tx_tables = 0, so it fixes the original problem.

Looks good to me besides the casting I mentioned.

**Review by seelabs — APPROVED** [review]
:+1: Thanks for changing the interface to use `-number` instead of `~number`.

File: src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp (hunk 5 above)
Comment (undertome) [reply src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp:269]: The compiler actually doesn't actually invoke the `bool` operator without the cast. Changed to `static_cast` instead.

File: src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp (hunk 6 above)
Comment (undertome) [reply src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp:279]: Changed to `static_cast`

File: src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp (hunk 2 above)
Comment (undertome) [reply src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp:1355]: Done

**Comment by undertome** [comment]
> There are a lot of places where we have to check the value of `useTxTables_`. It's seems like this might be a little error prone. What do you think of creating two separate classes, one for when `useTxTables_` is true, and one for when it's false? The class used in the false case would just return an empty optional for all of the functions that query the tx tables.

I like that approach in certain scenarios. For instance I used it with the `ShardArchiveHandler` where a derived type is instantiated when the handler is starting in recovery mode. But I think its a bit heavy handed to create a separate class with separate functions just for this single `bool`.

File: src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp (hunk 4 above)
Comment (undertome) [reply src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp:473]: Ended up removing them when addressing CJ's comment about returning early when transaction tables are disabled. Done

File: src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp [hunk 7]
```diff
@@ -266,7 +266,7 @@ class RelationalDBInterfaceSqliteImp : public RelationalDBInterfaceSqlite
     bool
     existsLedger()
     {
-        return (bool)lgrdb_;
+        return static_cast<bool>(lgrdb_);
```
Comment (mtrippled) [inline src/ripple/app/rdb/backend/RelationalDBInterfaceSqlite.cpp:269]: https://www.cplusplus.com/reference/memory/unique_ptr/operator%20bool/
The bool() operator in that object is called that does exactly what you want.

**Review by mtrippled — APPROVED** [review]
LGTM :+1:

**Comment by undertome** [comment]
> LGTM +1

Ran the test again - it checks out.

**Review by cjcobb23 — APPROVED** [review]
LGTM

**Pull request closed without merging**